# Running Ollama as a server with public url using ngrok to use it in mini-RAG App.

In [1]:
!curl -fsSL https://ollama.com/install.sh | sh

>>> Installing ollama to /usr/local
>>> Downloading Linux amd64 bundle
############################################################################################# 100.0%
>>> Creating ollama user...
>>> Adding ollama user to video group...
>>> Adding current user to ollama group...
>>> Creating ollama systemd service...
>>> The Ollama API is now available at 127.0.0.1:11434.
>>> Install complete. Run "ollama" from the command line.


In [2]:
ollama_model_id = 'deepseek-r1:7b'
# !pkill -f ollama >> to kill the process of ollama

**This two commands used to start the ollama server in the background and wait 5 seconds and then check the output of this cell in the nohup.out file but we need to make it listen to public requests not only the localhost**

```
!nohup ollama serve &
!sleep 5 && tail /content/nohup.out
```
#### so we can change the ollama_host and ollama_origin from os or in the command

**The default values for OLLAMA_HOST is 127.0.0.1:11434 and OLLAMA_ORIGIN is 127.0.0.1**

**We can change the host to be public 0.0.0.0 and the origin to serve for all requests**

In [ ]:
# !nohup bash -c "OLLAMA_HOST=0.0.0.0:4000 OLLAMA_ORIGIN=* ollama serve" &
# !sleep 5 && tail /content/nohup.out

In [3]:
import os
os.environ['OLLAMA_HOST'] = '0.0.0.0:4000'
os.environ['OLLAMA_ORIGIN'] = '*'

In [4]:
!nohup ollama serve &
!sleep 5 && tail /content/nohup.out

nohup: appending output to 'nohup.out'
[GIN-debug] GET    /                         --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func1 (5 handlers)
[GIN-debug] GET    /api/tags                 --> github.com/ollama/ollama/server.(*Server).ListHandler-fm (5 handlers)
[GIN-debug] GET    /api/version              --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func2 (5 handlers)
[GIN-debug] HEAD   /                         --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func1 (5 handlers)
[GIN-debug] HEAD   /api/tags                 --> github.com/ollama/ollama/server.(*Server).ListHandler-fm (5 handlers)
[GIN-debug] HEAD   /api/version              --> github.com/ollama/ollama/server.(*Server).GenerateRoutes.func2 (5 handlers)
time=2025-02-11T16:41:23.502Z level=INFO source=routes.go:1238 msg="Listening on [::]:4000 (version 0.5.7)"
time=2025-02-11T16:41:23.503Z level=INFO source=routes.go:1267 msg="Dynamic LLM libraries" runners="[cpu cpu_avx cp

In [5]:
# !OLLAMA_HOST=0.0.0.0:4000 OLLAMA_ORIGIN=* ollama pull {ollama_model_id}

!ollama pull {ollama_model_id}

pulling manifest ⠋ pulling manifest ⠙ pulling manifest ⠹ pulling manifest ⠸ pulling manifest ⠼ pulling manifest ⠴ pulling manifest ⠦ pulling manifest 
pulling 96c415656d37...   0% ▕▏    0 B/4.7 GB                  pulling manifest 
pulling 96c415656d37...   0% ▕▏    0 B/4.7 GB                  pulling manifest 
pulling 96c415656d37...   0% ▕▏ 1.8 MB/4.7 GB                  pulling manifest 
pulling 96c415656d37...   1% ▕▏  60 MB/4.7 GB                  pulling manifest 
pulling 96c415656d37...   3% ▕▏ 123 MB/4.7 GB                  pulling manifest 
pulling 96c415656d37...   3% ▕▏ 146 MB/4.7 GB                  pulling manifest 
pulling 96c415656d37...   4% ▕▏ 171 MB/4.7 GB                  pulling manifest 
pulling 96c415656d37...   4% ▕▏ 184 MB/4.7 GB                  pulling manifest 
pulling 96c415656d37...   5% ▕▏ 214 MB/4.7 GB                  pulling manifest 
pulling 96c415656d37...   5% ▕▏ 249 MB/4.7 GB                  pulling manifest 
pulling 96c415656d37...   6% ▕▏ 288 MB/

In [4]:
# !nohup bash -c "OLLAMA_HOST=0.0.0.0:4000 OLLAMA_ORIGIN=* ollama run {ollama_model_id}" &
# !sleep 5 && tail /content/nohup.out

nohup: appending output to 'nohup.out'
pulling manifest ⠋ pulling manifest ⠙ pulling manifest ⠹ pulling manifest ⠸ pulling manifest ⠼ pulling manifest ⠴ pulling manifest ⠦ pulling manifest ⠧ pulling manifest ⠇ [GIN] 2025/02/11 - 15:36:43 | 200 |  941.434771ms |       127.0.0.1 | POST     "/api/pull"
pulling manifest 
pulling 96c415656d37... 100% ▕████████████████▏ 4.7 GB                         
pulling 369ca498f347... 100% ▕████████████████▏  387 B                         
pulling 6e4c38e1172f... 100% ▕████████████████▏ 1.1 KB                         
pulling f4d24e9138dd... 100% ▕████████████████▏  148 B                         
pulling 40fb844194b2... 100% ▕████████████████▏  487 B                         
verifying sha256 digest 
writing manifest 
success 


In [7]:
!nohup ollama run {ollama_model_id} &
!sleep 5 && tail /content/nohup.out

nohup: appending output to 'nohup.out'
llm_load_print_meta: FIM MID token    = 151660 '<|fim_middle|>'
llm_load_print_meta: FIM PAD token    = 151662 '<|fim_pad|>'
llm_load_print_meta: FIM REP token    = 151663 '<|repo_name|>'
llm_load_print_meta: FIM SEP token    = 151664 '<|file_sep|>'
llm_load_print_meta: EOG token        = 151643 '<｜end▁of▁sentence｜>'
llm_load_print_meta: EOG token        = 151662 '<|fim_pad|>'
llm_load_print_meta: EOG token        = 151663 '<|repo_name|>'
llm_load_print_meta: EOG token        = 151664 '<|file_sep|>'
llm_load_print_meta: max token length = 256
⠇ ⠋ ⠙ ⠙ ⠸ ⠼ ⠴ ⠦ ⠦ ⠇ ⠏ ⠋ ⠙ ⠹ ⠸ ⠸ ⠴ ⠴ ⠧ ⠇ ⠏ ⠋ ⠋ ⠙ ⠹ ⠼ ⠴ ⠦ ⠧ ⠇ ⠇ ⠋ 

In [ ]:
%%bash

curl http://localhost:4000/api/chat -d '{
  "model": "deepseek-r1:7b",
  "messages": [
    { "role": "user", "content": "ما هي عاصمه مصر؟؟" }
  ]
}'

### Use ngrok to generate public url

In [10]:
!pip install pyngrok==7.2.3

In [13]:
from google.colab import userdata
from pyngrok import ngrok, conf

ngrok_auth = userdata.get('ngrok_auth')

conf.get_default().auth_token = ngrok_auth

port = "4000"

public_url = ngrok.connect(port).public_url
print(public_url)

https://bb3f-34-118-242-91.ngrok-free.app


### Use this public url to openai base url to in the .env file to be like that
**OPENAI_URL_BASE="[PUBLIC_URL]/v1"**

replace `[PUBLIC_URL]` with the output of the prev cell

### Change also the GENERATION_MODEL_ID from .env file to be the same as ollama_model_id from this notebook

In [14]:
%%bash

curl https://bb3f-34-118-242-91.ngrok-free.app/api/chat -d '{
  "model": "deepseek-r1:7b",
  "messages": [
    { "role": "user", "content": "ما هي عاصمه مصر؟؟" }
  ]
}'

{"model":"deepseek-r1:7b","created_at":"2025-02-11T16:47:13.580424073Z","message":{"role":"assistant","content":"\u003cthink\u003e"},"done":false}
{"model":"deepseek-r1:7b","created_at":"2025-02-11T16:47:13.620036027Z","message":{"role":"assistant","content":"\n\n"},"done":false}
{"model":"deepseek-r1:7b","created_at":"2025-02-11T16:47:13.660125051Z","message":{"role":"assistant","content":"\u003c/think\u003e"},"done":false}
{"model":"deepseek-r1:7b","created_at":"2025-02-11T16:47:13.700199961Z","message":{"role":"assistant","content":"\n\n"},"done":false}
{"model":"deepseek-r1:7b","created_at":"2025-02-11T16:47:13.738408319Z","message":{"role":"assistant","content":"ع"},"done":false}
{"model":"deepseek-r1:7b","created_at":"2025-02-11T16:47:13.767157462Z","message":{"role":"assistant","content":"اص"},"done":false}
{"model":"deepseek-r1:7b","created_at":"2025-02-11T16:47:13.795913963Z","message":{"role":"assistant","content":"مة"},"done":false}
{"model":"deepseek-r1:7b","created_at":"20

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 28516    0 28395  100   121   4543     19  0:00:06  0:00:06 --:--:--  4823
